File: src/day15/Day15.kt
package day15

import java.io.File

const val FILENAME = "src/day15/Input.txt"

val starting = File(FILENAME).readText().split(",").map { it.toInt() }.toTypedArray()

fun part01(limit: Int = 2020): Int {
    val prevOccurrence = IntArray(limit)
    //val prevOccurrence = mutableMapOf<Int, Int>()
    var turn = 0
    for (i in starting) prevOccurrence[i] = ++turn

    var curr = starting[starting.size - 1]
    var next = 0
    while (turn < limit) {
        turn++
        curr = next
        next = if (prevOccurrence[curr] == 0) 0 else turn - prevOccurrence[curr] // Uncomment when using array
        //next = if (!prevOccurrence.containsKey(curr)) 0 else turn - prevOccurrence[curr]!! // Uncomment when using map
        prevOccurrence[curr] = turn
    }

    return curr
}

fun part02() = part01(30_000_000)

fun main() {
    println("Advent of Code 2020, Day 15")
    println(part01())
    println(part02())
}

File: day15/day15.go
package day15

import (
	"advent-of-code-2020/utils"
	"strings"
)

const filename = "./inputs/day15.txt"

var starting = utils.MapToInt(strings.Split(utils.ReadString(filename), ","))

func numberSpoken(limit int) int {
	prevOccurrence := make([]int, limit)
	// prevOccurrence := make(map[int]int)
	turn := 0
	for _, i := range starting {
		turn++
		prevOccurrence[i] = turn
	}

	curr := starting[len(starting)-1]
	next := 0
	for turn < limit {
		turn++
		curr = next
		next = 0
		// Uncomment when using array
		if prevOccurrence[curr] != 0 {
			next = turn - prevOccurrence[curr]
		}
		// Uncomment when using map
		// if value, ok := prevOccurrence[curr]; ok {
		// 	next = turn - value
		// }
		prevOccurrence[curr] = turn
	}

	return curr
}

func part01() int {
	return numberSpoken(2020)
}

func part02() int {
	return numberSpoken(30_000_000)
}

func Main() (int, func() int, func() int) {
	return 15, part01, part02
}
